nexus/src/app/sagas/snapshot_create.rs

```diff
@@ -697,6 +1051,76 @@ async fn ssc_send_snapshot_request_undo(
     Ok(())
 }
 
+async fn ssc_call_pantry_detach_for_disk(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let log = sagactx.user_data().log();
+    let params = sagactx.saga_params::<Params>()?;
+
+    let pantry_address = sagactx.lookup::<SocketAddrV6>("pantry_address")?;
+
+    let endpoint = format!("http://{}", pantry_address);
+
+    info!(
+        log,
+        "sending detach for disk {} to endpoint {}", params.disk_id, endpoint,
+    );
+
+    let client = crucible_pantry_client::Client::new(&endpoint);
+
+    retry_until_known_result!(log, {
+        client.detach(&params.disk_id.to_string())
+    });
+
+    Ok(())
+}
+
+async fn ssc_detach_disk_from_pantry(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    let (.., authz_disk, db_disk) =
+        LookupPath::new(&opctx, &osagactx.datastore())
+            .disk_id(params.disk_id)
+            .fetch_for(authz::Action::Modify)
+            .await
+            .map_err(ActionError::action_failed)?;
```

Comment:
I think it's possible that this node might cause a data corruption in a case where we are concurrently snapshotting the same disk.

- Suppose we have a disk, D.
- Suppose we are executing two sagas on different nexus nodes to snapshot the disk D. 
- Saga 1 executes *almost* entirely, getting through this node, then crashing. This leaves the disk in state: "Detached", but `ssc_detach_disk_from_pantry` is not yet marked as done.
- Saga 2 starts, and is mid-way through the snapshot operation. The state of the disk is "Maintenance".
- Saga 1 restarts on a rebooted nexus node. It retries `ssc_detach_disk_from_pantry`.

With the current implementation, I think this would cause "saga 1" to pull the disk *out* of maintenance, even though it's actually being used by "saga 2".

I think the crux of this problem is that the generation accounting on the disk is lost, because the `db_disk` value is re-read from CRDB. As you mentioned [here](https://github.com/oxidecomputer/omicron/pull/1916#discussion_r1095108519), we would *want* to proceed with the "detach" operation only if the generation number is currently equal our expected value.

In other words, I think this is what we *want* to do during a snapshot saga using the pantry:
| Generation Number | State |
| --- | --- |
| N | Detached |
| N + 1 | Maintenance - but we only move it here if the prior `Generation` was N |
| N + 2 | Detached - but we only move it here if the prior `Generation` was N + 1 |

But I think this is what we're actually doing:
In other words, I think this is what we *want* to do during a snapshot saga using the pantry:
| Generation Number | State |
| --- | --- |
| N | Detached |
| N + 1 | Maintenance - but we only move it here if the prior `Generation` was N |
| X | Detached - but we only move it here if the prior `State` was `Maintenance`. But it might be in maintenance because of an operation made by a different saga! |

Reply:
I think you're right, yeah. 7ff4beb4 records the generation number N+1 (after setting to Maintenance), then looks it up later to verify that expectation before setting to Detached